feat(op): comprehensive metrics documentation (#37924)
DataDog/documentation: content/en/observability_pipelines/monitoring_and_troubleshooting/pipeline_usage_metrics.md

See [Getting Started with Tags][4] for more information on how to use tags to group metrics by specific pipelines, Workers, and components.

All metrics are tagged with the following:

`pipeline_id`
: The UUID of the pipeline.

`worker_uuid`
: The UUID of the Worker emitting the metric.

`op_worker_version`
: The version of the Worker emitting the metric.

`rc_version`
: The configuration version number, incremented each time the pipeline is updated.

`pipeline_name`
: The name of the pipeline when it was last deployed or updated. Available in Worker version 2.18 and later.

**Notes**:

> - Every Worker also runs an internal pipeline that collects the Worker's own telemetry (metrics and logs) and sends it to Datadog. The components in this internal pipeline have a `component_id` that starts with an underscore. To exclude them from your queries, filter with `!component_id:_*`.

Reviewer: ```suggestion
- Every Worker also runs an internal pipeline that collects the Worker's own telemetry (metrics and logs) and sends it to Datadog. The components in this internal pipeline have a `component_id` tag whose value starts with an underscore (`_`). To exclude these metrics from your queries, use `!component_id:_*`.
```

> **Note**: Metrics ending in `_total` report a count for each time interval, so their raw value does not increase monotonically.

Reviewer: Combining the notes.

> The components in this internal pipeline have a `component_id` that starts with an underscore. 

I don't quite understand that part. What are the `component_id:_*` metrics measuring?

```suggestion
**Notes**: 
- Every Worker also runs an internal pipeline that collects the Worker's own telemetry (metrics and logs) and sends it to Datadog. The components in this internal pipeline have a `component_id` that starts with an underscore. To exclude them from your queries, filter with `!component_id:_*`.
- Metrics ending in `_total` report a count for each time interval, so their raw value does not increase monotonically.
```

Author: They are reported for the internal pipeline components, which IDs are hardcoded as _xxxx.

## Host metrics

> These metrics describe the host running the Observability Pipelines Worker.

Reviewer: ```suggestion
These metrics provide information about the host running the Observability Pipelines Worker.
```

Uptime
: **Metric**: `pipelines.host.uptime`
: **Description:** The amount of time since the host was started, in seconds.

CPU time
: **Metric**: `pipelines.host.cpu_seconds_total`
: **Description:** The total CPU time consumed by the host, broken down by mode (user, system, idle, and so on) and CPU core.

Logical CPUs
: **Metric**: `pipelines.host.logical_cpus`
: **Description:** The number of logical CPU threads (hardware threads) available on the host.

Load average
: **Metric**: `pipelines.host.load1`, `pipelines.host.load5`, `pipelines.host.load15`

> : **Description:** The host's system load average over the last 1, 5, and 15 minutes. Load average is the number of processes that are running or waiting to run, and on Linux also includes processes blocked on uninterruptible I/O. Compare it against `pipelines.host.logical_cpus`: a value near the CPU count indicates full utilization, and a value above it indicates the host is oversubscribed. Not emitted on Windows.

Reviewer: ```suggestion
: **Description:** The host's system load average over the last 1, 5, and 15 minutes. Load average is the number of processes that are running or waiting to run, and on Linux also includes processes blocked on uninterruptible I/O. Compare the load average value against the `pipelines.host.logical_cpus` value: a load average value near the CPU count indicates full utilization, and a value above it indicates the host is oversubscribed. Not emitted on Workers running in Windows.
```

## Process metrics

> These metrics describe the Observability Pipelines Worker process.

Reviewer: Is runtime the correct word here?

```suggestion
These metrics provide information about the Observability Pipelines Worker's resource usage and runtime.
```

Author: `process` is more semantically accurate here. I'll apply the first suggestion

: **Description**: The byte size of events the component sends downstream.
: **Available for**: Sources, processors, and destinations.

Bytes in
: **Metric**: `pipelines.component_received_bytes_total`
: **Description**: The number of raw bytes read from the source's input, before any decoding or transformation.
: **Available for**: Sources.

Bytes out
: **Metric**: `pipelines.component_sent_bytes_total`

> : **Description**: The number of raw bytes written to the destination's output, after any encoding or transformation.

Reviewer: ```suggestion
: **Description**: The number of raw bytes written to the destination's output, after encoding and transformations.
```

: **Available for**: Destinations.

Events included
: **Metric**: `pipelines.included_events_total`

> : **Description**: The number of events that matched the processor's include clause and were processed. Events that do not match the include clause skip the processor and continue to the next component.

Reviewer: What's the include clause? Is that the filter query?

Author: yes. rephrased it

: **Available for**: Processors.

Event bytes included
: **Metric**: `pipelines.included_event_bytes_total`

> : **Description**: The byte size of events that matched the processor's include clause and were processed.

Reviewer: Same question as above

: **Description**: The number of requests that timed out for sources that send events to the Worker in batches using HTTP requests.
: **Available for**: HTTP-based sources that have a configured timeout, such as the Datadog Agent.

> Source lag time

Reviewer: Is it possible to give an example of a high value? Or does that vary?

Author: customer's judgement would be best here. I would say that a few seconds of delay sounds quite high, but I don't have studied this enough to be confident that this is always true.

: **Metric**: `pipelines.source_lag_time_seconds`
: **Description**: The difference, in seconds, between an event's own timestamp and when the Worker received it. High values indicate stale or delayed data arriving at the pipeline.
: **Available for**: Sources.

Utilization
: **Metric**: `pipelines.utilization`

> : **Description**: The component's activity. A value of `0` indicates an idle component that is waiting for input. A value close to `1` indicates a component that is never idle, which means that the component is likely a bottleneck in the processing topology that is creating backpressure, which might cause events to be dropped.

Reviewer: ```suggestion
: **Description**: The component's activity. A value of `0` indicates an idle component that is waiting for input. A value close to `1` indicates a component that is never idle, meaning that the component is likely a bottleneck in the processing topology that is creating backpressure. This might cause events to be dropped.
```

: **Available for**: Processors and destinations.

CPU usage
: **Metric**: `pipelines.component_cpu_usage_ns_total`
: **Description**: The CPU time consumed by a component, in nanoseconds. Use this metric to attribute CPU cost to individual processors. Available in Worker version 2.18 and later.

> : **Available for**: The Remap (VRL), Sensitive Data Scanner, Grok Parser, Parse JSON, Parse XML, OCSF Mapper, Enrichment Table, Reduce, Dedupe, Split Array, and Throttle log processors, and the Aggregate and Tag Cardinality Limit metrics processors.

Reviewer: Is the Remap processor the Custom Processor or the Edit Fields processor? Do we need to include the `(VRL)` part?

```suggestion
: **Available for**: The Custom Processor (VRL), Sensitive Data Scanner, Grok Parser, Parse JSON, Parse XML, Remap to OCSF, Enrichment Table, Reduce, Dedupe, Split Array, and Throttle log processors, and the Aggregate and Tag Cardinality Limit metrics processors.
```

Author: yes, it is. unfortunately the internal terminology (component type) is `remap_vrl`, but we could drop the VRL part here.

Author: done

: **Available for**: Sources.

Send batch latency
: **Metric**: `pipelines.source_send_batch_latency_seconds`
: **Description**: The time it takes for the source to send a batch, which can contain multiple event chunks, to the next component. Available in Worker version 2.16 and later.
: **Available for**: Sources.

## HTTP server metrics

These metrics are emitted by sources that receive data over HTTP, such as the Datadog Agent, HTTP/S Server, Splunk HEC, and OpenTelemetry sources.

- Use the `component_id` tag to filter or group by individual components.
- Use the `component_type` tag to filter or group by the source type.

`pipelines.http_server_requests_received_total`
: **Description**: The number of HTTP requests received.
: **Metric type**: count

`pipelines.http_server_responses_sent_total`
: **Description**: The number of HTTP responses sent.
: **Metric type**: count

`pipelines.http_server_handler_duration_seconds`
: **Description**: The time spent handling an HTTP request.
: **Metric type**: distribution

## HTTP client metrics

> These metrics are emitted by destinations that send data over HTTP, such as Datadog Logs, Datadog Metrics, Elasticsearch, OpenSearch, New Relic, Splunk HEC, Microsoft Sentinel, SentinelOne, CrowdStrike NG-SIEM, Google SecOps, and the HTTP Client destination. **Note**: AWS-based destinations (such as Amazon S3, Amazon OpenSearch, and Amazon Security Lake) do not emit these metrics.

Reviewer: Can we list out the destinations? I think it'll be easier to read. For example:

- Datadog Logs
- Datadog Metrics]
- Elasticsearch
- ...

- Use the `component_type` tag to filter or group by the destination type.

`pipelines.http_client_requests_sent_total`
: **Description**: The number of HTTP requests sent, tagged by request method.
: **Metric type**: count

`pipelines.http_client_responses_total`
: **Description**: The number of HTTP responses received, tagged by response status.
: **Metric type**: count

`pipelines.http_client_errors_total`
: **Description**: The number of HTTP client errors, tagged by error kind.
: **Metric type**: count

`pipelines.http_client_rtt_seconds`
: **Description**: The round-trip time, in seconds, for HTTP requests, from when the request is sent to when the final response or error is received.
: **Metric type**: distribution

`pipelines.http_client_response_rtt_seconds`
: **Description**: The round-trip time, in seconds, of HTTP requests, tagged by response status.
: **Metric type**: distribution

`pipelines.http_client_error_rtt_seconds`
: **Description**: The round-trip time, in seconds, of HTTP requests that resulted in an error, tagged by error kind.
: **Metric type**: distribution

## Adaptive concurrency metrics

> These metrics describe the adaptive concurrency controller, which automatically tunes how many in-flight HTTP requests a destination allows based on observed response times. They are emitted by destinations that send data over HTTP, including AWS-based destinations.

Reviewer: ```suggestion
These metrics provide information about the adaptive concurrency controller, which automatically tunes how many in-flight HTTP requests a destination allows based on observed response times. They are emitted by destinations that send data over HTTP, including AWS-based destinations.
```

- Use the `component_id` tag to filter or group by individual components.
- Use the `component_type` tag to filter or group by the destination type.

`pipelines.active_endpoints`

> : **Description**: The number of sink endpoints that are marked healthy.

Reviewer: ```suggestion
: **Description**: The number of destination endpoints that are marked healthy.
```

: **Metric type**: gauge

`pipelines.adaptive_concurrency_limit`

> : **Description**: The concurrency limit for HTTP requests to this destination, automatically adjusted by the adaptive concurrency controller based on observed response times.

Reviewer: Does `observed` mean anything specific here? If not, suggest removing it.

```suggestion
: **Description**: The concurrency limit for HTTP requests to this destination, automatically adjusted by the adaptive concurrency controller based on response times.
```

: **Metric type**: distribution

`pipelines.adaptive_concurrency_in_flight`
: **Description**: The number of HTTP requests in flight to a destination, compared against the adaptive concurrency limit to determine when to throttle.
: **Metric type**: distribution

`pipelines.adaptive_concurrency_reached_limit`

> : **Description**: Whether the adaptive concurrency controller reached its computed limit (1) or not (0) during the last measurement interval.

Reviewer: Does this mean the metric emits either a `1` or `0`?

```suggestion
: **Description**: Whether the adaptive concurrency controller reached its computed limit (`1`) or not (`0`) during the last measurement interval.
```

Author: yes

: **Metric type**: distribution

`pipelines.adaptive_concurrency_back_pressure`

> : **Description**: Whether the adaptive concurrency controller detected back pressure (1) or not (0) during the last measurement interval.

Reviewer: Same question as above

```suggestion
: **Description**: Whether the adaptive concurrency controller detected back pressure (`1`) or not (`0`) during the last measurement interval.
```
